File: sentinelai-backend/app/schemas/mitre.py
from datetime import datetime
from typing import Any

from pydantic import Field

from app.schemas.base import BaseSchema


class MitreTechniqueResponse(BaseSchema):
    id: str
    technique_id: str
    name: str
    description: str | None = None
    tactic: str
    tactic_id: str | None = None
    platform: list[str] | None = None
    permissions_required: list[str] | None = None
    detection: str | None = None
    is_subtechnique: bool = False
    parent_technique_id: str | None = None
    severity: str = "medium"
    score: float = 1.0
    mitre_version: str = "15.1"
    detection_rules: list[str] | None = None
    ioc_indicators: list[str] | None = None
    kill_chain_phase: str | None = None
    data_sources: list[str] | None = None
    url: str | None = None
    created_at: datetime
    updated_at: datetime


class MitreMappingResponse(BaseSchema):
    id: str
    technique_id: str
    mapped_type: str
    mapped_id: str
    mapped_name: str | None = None
    confidence: float
    source: str
    context: str | None = None
    mapped_at: datetime
    extra_data: dict[str, Any] | None = Field(None, validation_alias="metadata")
    created_at: datetime


class CoverageStatisticResponse(BaseSchema):
    tactic: str
    total_techniques: int
    mapped_techniques: int
    coverage_percent: float
    total_detections: int
    mapped_detections: int
    avg_confidence: float
    calculated_at: datetime


class MitreCoverageResponse(BaseSchema):
    overall_coverage: float
    total_techniques: int
    total_mapped: int
    total_detections: int
    by_tactic: list[CoverageStatisticResponse]
    top_techniques: list[dict[str, Any]]
    top_tactics: list[dict[str, Any]]
    most_triggered: list[dict[str, Any]]


class MitreMapRequest(BaseSchema):
    mapped_type: str
    mapped_id: str
    mapped_name: str | None = None
    context: str | None = None


class MitreMapResponse(BaseSchema):
    mappings: list[MitreMappingResponse]
    new_mappings: int
    confidence_avg: float


class MitreTechniqueDetail(BaseSchema):
    technique: MitreTechniqueResponse
    mappings: list[MitreMappingResponse]
    mapped_count: int
    detection_coverage: float
    related_techniques: list[MitreTechniqueResponse]

File: sentinelai-frontend/src/types/mitre.ts
export interface MitreTechnique {
  id: string;
  technique_id: string;
  name: string;
  description: string | null;
  tactic: string;
  tactic_id: string | null;
  platform: string[] | null;
  permissions_required: string[] | null;
  detection: string | null;
  is_subtechnique: boolean;
  parent_technique_id: string | null;
  severity: string;
  score: number;
  mitre_version: string;
  detection_rules: string[] | null;
  ioc_indicators: string[] | null;
  kill_chain_phase: string | null;
  data_sources: string[] | null;
  url: string | null;
  created_at: string;
  updated_at: string;
}

export interface MitreMapping {
  id: string;
  technique_id: string;
  mapped_type: string;
  mapped_id: string;
  mapped_name: string | null;
  confidence: number;
  source: string;
  context: string | null;
  mapped_at: string;
  created_at: string;
}

export interface CoverageStatistic {
  tactic: string;
  total_techniques: number;
  mapped_techniques: number;
  coverage_percent: number;
  total_detections: number;
  mapped_detections: number;
  avg_confidence: number;
  calculated_at: string;
}

export interface MitreCoverage {
  overall_coverage: number;
  total_techniques: number;
  total_mapped: number;
  total_detections: number;
  by_tactic: CoverageStatistic[];
  top_techniques: { technique_id: string; name: string; count: number }[];
  top_tactics: CoverageStatistic[];
  most_triggered: { technique_id: string; name: string; tactic: string; count: number }[];
}

export interface MitreTechniqueDetail {
  technique: MitreTechnique;
  mappings: MitreMapping[];
  mapped_count: number;
  detection_coverage: number;
  related_techniques: MitreTechnique[];
}

export const TACTIC_COLORS: Record<string, string> = {
  'Initial Access': '#ef4444',
  Execution: '#f97316',
  Persistence: '#eab308',
  'Privilege Escalation': '#22c55e',
  'Defense Evasion': '#14b8a6',
  'Credential Access': '#3b82f6',
  Discovery: '#6366f1',
  'Lateral Movement': '#8b5cf6',
  Collection: '#ec4899',
  'Command and Control': '#f43f5e',
  Exfiltration: '#a855f7',
  Impact: '#dc2626',
};

export const MITRE_SEVERITY_COLORS: Record<string, string> = {
  critical: 'text-red-400 bg-red-400/10 border-red-400/30',
  high: 'text-orange-400 bg-orange-400/10 border-orange-400/30',
  medium: 'text-yellow-400 bg-yellow-400/10 border-yellow-400/30',
  low: 'text-green-400 bg-green-400/10 border-green-400/30',
};
